librosa/core/pitch.py

```diff
@@ -335,3 +338,281 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+    parabola_a = (yin_frames[:-2, :] + yin_frames[2:, :] - 2*yin_frames[1:-1, :]) / 2
+    parabola_b = (yin_frames[2:, :] - yin_frames[:-2, :]) / 2
+    parabolic_shifts = -parabola_b / (2*parabola_a + util.tiny(parabola_a))
+    parabolic_shifts[np.abs(parabolic_shifts) > 1] = 0
+    yin_frames = yin_frames[1:-1, :]
+
+    # Find Yin candidates and probabilities.
+    # The implementation here follows the official pYIN software which
+    # differs from the method described in the paper.
+    # 1. Define the prior over the thresholds.
+    n_thresholds = 100  # number of thresholds for peak estimation.
+    thresholds = np.linspace(0, 1, n_thresholds + 1)
+    beta_cdf = scipy.stats.beta.cdf(thresholds, 2, 18)
+    beta_probs = np.diff(beta_cdf)
+
+    yin_probs = np.zeros_like(yin_frames)
+    for i, yin_frame in enumerate(yin_frames.T):
+        # 2. For each frame find the troughs.
+        is_trough = np.logical_and(
+            yin_frame[1:-1] < yin_frame[:-2],
+            yin_frame[1:-1] < yin_frame[2:])
+        trough_index, = np.nonzero(is_trough)
+
+        if len(trough_index) == 0:
+            continue
+
+        # 3. Find the troughs below each threshold.
+        trough_heights = yin_frame[trough_index + 1]
+        trough_thresholds = trough_heights[:, None] < thresholds[None, 1:]
+
+        # 4. Define the prior over the troughs.
+        # Smaller periods are weighted more.
+        trough_positions = np.cumsum(trough_thresholds, axis=0) - 1
+        n_troughs = np.count_nonzero(trough_thresholds, axis=0)
+        trough_prior = scipy.stats.boltzmann.pmf(trough_positions, 10, n_troughs)
+        trough_prior[~trough_thresholds] = 0
+
+        # 5. For each threshold add probability to global minimum if no trough is below threshold,
+        # else add probability to each trough below threshold biased by prior.
+        probs = np.sum(trough_prior * beta_probs, axis=1)
+        global_min = np.argmin(trough_heights)
+        n_thresholds_below_min = np.count_nonzero(~trough_thresholds[global_min, :])
+        probs[global_min] += 0.01 * np.sum(beta_probs[:n_thresholds_below_min])
+
+        yin_probs[trough_index + 1, i] = probs
+
+    yin_period, frame_index = np.nonzero(yin_probs)
+
+    # Refine peak by parabolic interpolation.
+    period_candidates = min_period + yin_period
+    period_candidates = period_candidates + parabolic_shifts[yin_period, frame_index]
+    f0_candidates = sr / period_candidates
+
+    n_bins_per_semitone = 10  # bins for discretizing the pitch space between fmin and fmax.
+    n_pitch_bins = int(np.log2(fmax / fmin) * 12 * n_bins_per_semitone)
+
+    # Construct transition matrix.
+    transition_width = 10 * (n_bins_per_semitone // 2) + 1
+    transition = sequence.transition_local(
+        n_pitch_bins, transition_width, window='triangle', wrap=False)
+    transition = np.block(
+        [[0.99 * transition, 0.01 * transition],
+         [0.01 * transition, 0.99 * transition]])
```

Comment:
transition built twice?

Reply:
sorry, misread that the first time -- the second line depends on the first.  This is okay, but deserves a comment to explain what's going on.

Reply:
also the 0.99 should be a user-settable parameter to control switching between voiced/unvoiced